### PR TITLE
[1/2] Implement node result and content routes

### PR DESCRIPTION
**Pull request by likawind**
## Describe your changes and why you are making these changes
This PR implements 4 routes regarding node results and contents. The LOC is large, but some files doesn't require extensive review as they are mostly copied.
* `/api/v2/workflow/{workflowID}/dag/{dagID}/node/artifact/{nodeID}/result/{nodeResultID}/content`: fetch contents (e.g. table) from an artf result. This is mostly copy from V1 `get_artifact_result` route.
* `/api/v2/workflow/{workflowID}/dag/{dagID}/node/artifact/{nodeID}/results`: Get artifact result's history. This is mostly copied from V1 `list_artifact_results`.
* `/api/v2/workflow/{workflowID}/dag/{dagID}/node/operator/{nodeID}/content` get operator contents. This is the same as `export function` but with much simplified implementations.
* `/api/v2/workflow/{workflowID}/result/{dagResultID}/nodes/results` to fetch all node results of a given DAG result.

Additionally, we added a `functional_primitives` library to simplify map / slice manipulation. This is inspired by Hack language: https://docs.hhvm.com/hsl/reference/function/HH.Lib.Dict.associate/
## Related issue number (if any)
ENG-2745

## Loom demo (if any)
Compile + integration tests should be sufficient

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/cmd/server/handler/v2/node_artifact_result_content_get.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+		AqContext:    aqContext,
+		workflowID:   workflowID,
+		dagID:        dagID,
+		nodeID:       nodeID,
+		nodeResultID: nodeResultID,
+	}, http.StatusOK, nil
+}
+
+func (h *NodeArtifactResultContentGetHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
+	args := interfaceArgs.(*nodeResultGetArgs)
```
Comment (likawind) [inline src/golang/cmd/server/handler/v2/node_artifact_result_content_get.go:107]: Not much need to review here, the implementation is copied from current `get_artifact_result` route. We will add necessary fix once we start integrate with UI.

File: src/golang/cmd/server/handler/v2/node_artifact_result_content_get.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	return &nodeResultGetResponse{IsDownsampled: isDownsampled, Content: data}, http.StatusOK, nil
+}
+
+// This custom implementation of SendResponse constructs a multipart form response with two fields:
+// 1: "metadata" contains a json serialized blob of artifact result metadata.
+// 2: "data" contains the artifact result data blob generated the serialization method
+// specified in the metadata field.
+func (*NodeArtifactResultContentGetHandler) SendResponse(w http.ResponseWriter, interfaceResp interface{}) {
```
Comment (likawind) [inline src/golang/cmd/server/handler/v2/node_artifact_result_content_get.go:176]: similarly, this is copied from `get_artifact_result` route.

File: src/golang/cmd/server/handler/v2/node_artifact_results_get.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+func (*NodeArtifactResultsGetHandler) Name() string {
+	return "NodeArtifactResultsGet"
+}
+
+func (h *NodeArtifactResultsGetHandler) Prepare(r *http.Request) (interface{}, int, error) {
+	return h.nodeGetHandler.Prepare(r)
+}
+
+func (h *NodeArtifactResultsGetHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
+	args := interfaceArgs.(*nodeGetArgs)
```
Comment (likawind) [inline src/golang/cmd/server/handler/v2/node_artifact_results_get.go:56]: Not much need to review here. This is copied from current `list_artifact_results` route.

File: src/golang/cmd/server/handler/v2/node_operator_content_get.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+func (*NodeOperatorContentGetHandler) Name() string {
+	return "NodeOperatorContentGet"
+}
+
+func (h *NodeOperatorContentGetHandler) Prepare(r *http.Request) (interface{}, int, error) {
+	return h.nodeGetHandler.Prepare(r)
+}
+
+func (h *NodeOperatorContentGetHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
```
Comment (likawind) [inline src/golang/cmd/server/handler/v2/node_operator_content_get.go:53]: review will be helpful. This is similar to current `export function` route but some implementation is simpler.

File: src/golang/cmd/server/handler/v2/nodes_results_get.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		ctx,
+		[]uuid.UUID{args.dagResultID},
+		h.Database,
+	)
+	if err != nil {
+		return nil, http.StatusInternalServerError, errors.Wrap(err, "Unexpected error occurred when retrieving artifact results.")
+	}
+
+	dbArtifacts, err := h.ArtifactRepo.GetByDAG(ctx, dbDAG.ID, h.Database)
+	dbArtifactsByID := functional_map.FromValues(
```
Comment (likawind) [inline src/golang/cmd/server/handler/v2/nodes_results_get.go:121]: example use of functional library

File: src/golang/lib/response/node.go [hunk 6]
```diff
@@ -92,17 +127,55 @@ func NewNodesFromDBObjects(
 	operatorNodes []views.OperatorNode,
 	artifactNodes []views.ArtifactNode,
 ) *Nodes {
-	operators := make([]Operator, 0, len(operatorNodes))
-	artifacts := make([]Artifact, 0, len(artifactNodes))
-	for _, opNode := range operatorNodes {
-		operators = append(operators, *NewOperatorFromDBObject(&opNode))
+	return &Nodes{
+		Operators: functional_slice.Map(
```
Comment (likawind) [inline src/golang/lib/response/node.go:131]: example use of functional library

File: src/golang/lib/functional_primitives/functional_slice/slice.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package functional_slice
```
Comment (likawind) [inline src/golang/lib/functional_primitives/functional_slice/slice.go:1]: @saurav-c what do you suggest for naming? The main issue is `maps` and `slice` appear to be reserved. As a result I couldn't do `maps / slices` or `map/slice` :(

Comment (saurav-c) [reply src/golang/lib/functional_primitives/functional_slice/slice.go:1]: Maybe just `lib/functional` and then we can `map.go` and `slice.go` inside.

File: src/golang/lib/response/node.go [hunk 8]
```diff
@@ -91,17 +126,55 @@ func NewNodesFromDBObjects(
...
+
+				return *NewArtifactResultFromDBObject(&result, contentPtr)
+			},
+		),
 	}
 }
+
+type Content struct {
+	Name string `json:"name"`
+	Data []byte `json:"data"`
```
Comment (kenxu95) [inline src/golang/lib/response/node.go:179]: I'm confused as to where this is used and what it represents. Can you add some explanation here?

**Review by kenxu95 — APPROVED** [review]

File: src/golang/lib/functional_primitives/functional_slice/slice.go (hunk 7 above)
Comment (likawind) [reply src/golang/lib/functional_primitives/functional_slice/slice.go]: Actually somehow `maps` and `slices` working, and we can proceed with these names

**Pull request merged**